library/core/src/option.rs

```diff
@@ -733,6 +734,86 @@ impl<T> Option<T> {
         }
     }
 
+    const fn get_some_offset() -> isize {
+        if mem::size_of::<Option<T>>() == mem::size_of::<T>() {
+            // niche optimization means the `T` is always stored at the same position as the Option.
+            0
+        } else {
+            assert!(mem::size_of::<Option<T>>() == mem::size_of::<Option<mem::MaybeUninit<T>>>());
+            let some_uninit = Some(mem::MaybeUninit::<T>::uninit());
+            // SAFETY: This gets the byte offset of the `Some(_)` value following the fact that
+            // niche optimization is not active, and thus Option<T> and Option<MaybeUninit<t>> share
+            // the same layout, which we have asserted above.
```

Comment:
This is not correct in general. There's no reason that `Option<u8>` can't be organized like `[discriminant, value]` while `Option<MaybeUninit<u8>>` is organized like `[value, discriminant]`. As far as I know that is not possible in our layout algorithm today, but non-std code should not rely on that. Within core it's probably fine to have in principle, although I am concerned about what happens when someone does change the layout algorithm so that this actually becomes wrong. At the very least, we need to make sure that they know that they are about to break this and don't silently introduce a bug instead - I don't know the best way to do that, maybe a well-placed comment somewhere in our layout logic suffices?

Reply:
Hmm, this is a good point.  I don't remember if randomize-layout can randomize discriminant position.

Maybe it'd be better to just add an intrinsic?  A generic "enum payload offset" would be hard, but this would only need to work for `Option` (or "option-like", I guess) so it could make a whole bunch of simplifying assumptions in the implementation.  And it's a compile-time calculation, so hopefully all the backends could just call the same logic and emitting the constant would be easy.

Reply:
Sure, but that RFC doesn't include enums, and while it's possible that we'll get them in the near future, it's also completely possible that it won't happen for a while, so I don't think we should be working under the assumption that this is temporary. I do like Scott's idea of inserting an intrinsic

Reply:
I think it would be easy enough to just implement the extension behind a feature flag, instead of creating yet another intrinsic.

Reply:
Btw. this shouldn't block us; I can simply use a `match` in those cases. It will probably be just as good as what someone else would write, with the optimization only applying to niche-optimized types for now, but we can improve that later.

Reply:
@scottmcm I have reviewed the layout randomization implementation and can confirm that it only works on a variant by variant case, never touching the enum discriminant at all. Since our only data-bearing variant here has but one field, layout randomization won't hurt us until someone changes that.

I now see two possible approaches:

1. I'll leave it as is and add a test case that would break horrifically if the layout algorithm ever changed in an incompatible way.
2. Change the code to use a `match` (and thus pessimizing the code to the state of what a layman would write) for the non-niche-optimized case until we have an extended `offset_of!` macro.

I'm fine with both options, leaning slightly towards the first.